### PR TITLE
online nsrt learning

### PR DESCRIPTION
**Pull request by tomsilver**
@ronuchit can you comment on the design here before I add tests?

### CODE REVIEW

File: src/approaches/online_nsrt_learning_approach.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
+"""An NSRT learning approach that collects and learns from data online.
+
+Example command:
+    python src/main.py --approach online_nsrt_learning --seed 0 \
+        --env cover \
+        --max_initial_demos 1 \
+        --num_train_tasks 1000 \
+        --num_test_tasks 10 \
+        --max_num_steps_interaction_request 10 \
```
Comment (ronuchit) [inline src/approaches/online_nsrt_learning_approach.py:9]: why is this so low?

File: src/approaches/online_nsrt_learning_approach.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        # explored at each time step is a hyperparameter. The train task
+        # is randomly selected.
+        explorer = self._get_explorer()
+        requests = []
+        for _ in range(CFG.online_nsrt_learning_tasks_per_request):
+            # Select a random task (with replacement).
+            task_idx = self._rng.choice(len(self._train_tasks))
+            task = self._train_tasks[task_idx]
+            # Set up the explorer policy and termination function.
+            policy, terminal = explorer.solve(task)
```
Comment (ronuchit) [inline src/approaches/online_nsrt_learning_approach.py:55]: rename to `termination_function`

**Review by ronuchit — COMMENTED** [review]
looks reasonable, but we'll see how things evolve when you add in GLIB. question: shouldn't we be unifying this with the stuff going on in `interactive_learning_approach.py`?

File: src/approaches/online_nsrt_learning_approach.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    def learn_from_offline_dataset(self, dataset: Dataset) -> None:
+        # Update the dataset with the offline data.
+        self._dataset = Dataset(dataset.trajectories)
+        super().learn_from_offline_dataset(dataset)
+
+    def learn_from_interaction_results(
+            self, results: Sequence[InteractionResult]) -> None:
+        # Add the new data to the cumulative dataset.
+        for result in results:
+            traj = LowLevelTrajectory(result.states, result.actions)
```
[the remark this answers is not in the document]
Comment (tomsilver) [reply src/approaches/online_nsrt_learning_approach.py:73]: good point. I'm looking through `git grep '.is_demo'` now. I know that predicate learning will filter this out. NSRT learning should be fine I think. we will have to come back for predicate learning...

File: src/approaches/online_nsrt_learning_approach.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        self._dataset = Dataset([])
+        self._online_learning_cycle = 0
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "online_nsrt_learning"
+
+    def get_interaction_requests(self) -> List[InteractionRequest]:
+        # Explore in the train tasks. The number of train tasks that are
+        # explored at each time step is a hyperparameter. The train task
```
[the remark this answers is not in the document]
Comment (tomsilver) [reply src/approaches/online_nsrt_learning_approach.py:46]: @&*#$*#@&2!

File: src/approaches/online_nsrt_learning_approach.py (hunk 1 above)
Comment (tomsilver) [reply src/approaches/online_nsrt_learning_approach.py:9]: just because each action is a pick or a place in cover

**Comment by tomsilver** [comment]
the new tests add about 0.75 seconds

**Comment by tomsilver** [comment]
>question: shouldn't we be unifying this with the stuff going on in interactive_learning_approach.py?
yeah but I would rather wait for Amber to get back before messing with that too much. I want her to be able to recognize her code lol. I will make an issue for the refactoring

**Comment by tomsilver** [comment]
@ronuchit let's merge this first and then I'll add GLIB?

**Comment by ronuchit** [comment]
i really think we should do that refactoring now (or not merge this) -- it's not going to change her code much unless i'm missing something. if you look at it, you basically re-implemented `create_random_interaction_strategy` here. all we need to do is pull that function out into a common place where both can use it (utils? or some new Explorer class in base_approach?). and then GLIB comes for free as well cause you can just use `create_glib_interaction_strategy`

**Comment by ronuchit** [comment]
do you really need an Explorer class? what's wrong with the way it's done in `interactive_learning_approach.py`?

**Pull request closed without merging**